src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -0,0 +1,176 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.graphanalysis;
+
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Scanner that will scan down forks when we hit parallel blocks.
+ * Think of it as the opposite of {@link DepthFirstScanner}:
+ *   - We visit every node exactly once, but walk through all parallel forks before resuming the main flow
+ *
+ * This is near-optimal in many cases, since it keeps minimal state information and explores parallel blocks first
+ * It is also very easy to make it branch/block-aware, since we have all the fork information at all times.
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public class ForkScanner extends AbstractFlowScanner {
+
+    /** These are the BlockStartNodes that begin parallel blocks
+     *  There will be one entry for every executing parallel branch in current flow
+     */
+    ArrayDeque<FlowNode> forkStarts = new ArrayDeque<FlowNode>();
+
+    /** FlowNode that will terminate the current parallel block */
+    FlowNode currentParallelStart = null;
+
+    /** How deep are we in parallel branches, if 0 we are linear */
+    protected int parallelDepth = 0;
+
+    @Override
+    protected void reset() {
+        if (_queue == null) {
+            _queue = new ArrayDeque<FlowNode>();
+        } else {
+            _queue.clear();
+        }
+        forkStarts.clear();
+        parallelDepth =0;
+        currentParallelStart = null;
+        _current = null;
+        _next = null;
+    }
+
+    @Override
+    protected void setHeads(@Nonnull Collection<FlowNode> heads) {
+        if (heads.size() > 1) {
+            throw new IllegalArgumentException("ForkedFlowScanner can't handle multiple head nodes yet");
```

Comment:
This is the final use case to handle.  I'm aware that it can't handle multiple heads yet: this should be the only remaining WIP component, but but it is quite complex to work with.